From 13beb56acdca59b70bad2a26f86a9c4dbac8f60b Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Thu, 22 Aug 2013 20:22:56 +0100
Subject: [PATCH 175/202] TV: Cache the last groupname found in
 ChannelGroup::GetChannelGroupNam

This significantly reduces dbase lookups during cursor movements ith the GuideGrid

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmythtv/channelgroup.cpp |   21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/mythtv/libs/libmythtv/channelgroup.cpp b/mythtv/libs/libmythtv/channelgroup.cpp
index ba2ea85..7fa255e 100644
--- a/mythtv/libs/libmythtv/channelgroup.cpp
+++ b/mythtv/libs/libmythtv/channelgroup.cpp
@@ -1,6 +1,9 @@
 // c++
 #include <algorithm>
 
+#include <QMutex>
+#include <QMutexLocker>
+
 // mythtv
 #include "mythlogging.h"
 #include "mythdb.h"
@@ -194,6 +197,11 @@ int ChannelGroup::GetNextChannelGroup(const ChannelGroupList &sorted, int grpid)
     return it->grpid;
 }
 
+static QMutex s_mutex;
+static int s_grpid = -1;
+static QString s_groupName;
+
+// static
 QString ChannelGroup::GetChannelGroupName(int grpid)
 {
     // All Channels
@@ -204,6 +212,11 @@ QString ChannelGroup::GetChannelGroupName(int grpid)
     if (grpid == 0)
         return "";
 
+    QMutexLocker lock(&s_mutex);
+    if (grpid == s_grpid)
+        return s_groupName;
+    lock.unlock();
+
     MSqlQuery query(MSqlQuery::InitCon());
     query.prepare("SELECT name FROM channelgroupnames WHERE grpid = :GROUPID");
     query.bindValue(":GROUPID", grpid);
@@ -211,7 +224,13 @@ QString ChannelGroup::GetChannelGroupName(int grpid)
     if (!query.exec())
         MythDB::DBError("ChannelGroup::GetChannelGroups", query);
     else if (query.next())
-        return query.value(0).toString();
+    {
+        QString groupName = query.value(0).toString();
+        lock.relock();
+        s_groupName = groupName;
+        s_grpid = grpid;
+        return groupName;
+    }
 
     return "";
 }
-- 
1.7.9.5

